From git@z Thu Jan  1 00:00:00 1970
Subject: [PATCH v1] bus: mhi: host: pci_generic: round up nr_irqs to a
 power of two
From: Javier Achirica <jachirica@gmail.com>
Date: Fri, 03 Apr 2026 12:30:03 +0200
Message-Id: <CACixm21q7b_diEx5COZxVZm9EhZ0hnakM_WBjEWcCsznfWeniw@mail.gmail.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 7bit

When an MHI device uses standard MSI, the PCI core requires the allocated
number of vectors to be a strict power of two but devices will only ask for the
irqs they need, so they might not be properly aligned.

Make sure a power-of-2 number of vectors is requested.

Signed-off-by: Javier Achirica <jachirica@gmail.com>

--- a/drivers/bus/mhi/host/pci_generic.c
+++ b/drivers/bus/mhi/host/pci_generic.c
@@ -1014,7 +1014,8 @@ static int mhi_pci_get_irqs(struct mhi_c
 	 */
 	mhi_cntrl->nr_irqs = 1 + mhi_cntrl_config->num_events;
 
-	nr_vectors = pci_alloc_irq_vectors(pdev, 1, mhi_cntrl->nr_irqs, PCI_IRQ_MSI);
+	nr_vectors = pci_alloc_irq_vectors(pdev, 1,
+			roundup_pow_of_two(mhi_cntrl->nr_irqs), PCI_IRQ_MSI);
 	if (nr_vectors < 0) {
 		dev_err(&pdev->dev, "Error allocating MSI vectors %d\n",
 			nr_vectors);
